src/peft/tuners/xlora/layer.py

```diff
@@ -0,0 +1,184 @@
+
+import torch
+import torch.nn as nn
+from torch import Tensor
+
+from peft.tuners import lora
+
+from .config import XLoraConfig
+
+
+class XLoraLayer:
+    """
+    A XLoraLayer wraps any LoraLayer and performs the XLora operation on the LoRA adaptors specified. Its primary API
+    is the forward method, which uses the scalings to execute the XLora algorithm.
+    """
+
+    def __init__(
+        self,
+        model: nn.Module,  # XLoraModel
+        target: lora.LoraLayer,
+        target_forward: Callable[..., Any],
+        layer_number: int,
+        config: XLoraConfig,
+    ) -> None:
+        self.model = model
+        self.target_forward = target_forward
+        self.target = target
+        self.layer_number = layer_number
+        self.config = config
+
+    @staticmethod
+    def apply_scalings_to_x(x: torch.Tensor, scalings_layer: torch.Tensor, adapter: int) -> torch.Tensor:
+        # scalings_layer = [batch_size, seq_len, n_classes]
+        scalings = scalings_layer[:, :, adapter].unsqueeze(-1)
+        # scalings_layer = [batch_size, seq_len, 1]
+        return x * scalings
+
+    def get_maybe_topk_scalings(self) -> torch.Tensor:
+        # XLora_scalings = [batch_size, seq_len, n_classes]
+        XLora_scalings: Tensor = self.scalings[:, :, self.layer_number, :]  # type: ignore
+
+        if self.config.top_k_lora is not None:
+            _, topk_indices = torch.topk(XLora_scalings, k=self.config.top_k_lora, dim=-1)
+
+            # Mask the topk to True, the rest to False
+            mask = torch.zeros_like(XLora_scalings, dtype=torch.bool)
+            mask.scatter_(-1, topk_indices, True)
+
+            XLora_scalings = XLora_scalings * mask.to(XLora_scalings.dtype)
+
+        if self.config.enable_softmax_topk:
+            nonzero_mask = XLora_scalings != 0
+            softmax_res_nonzero = torch.softmax(XLora_scalings[nonzero_mask], dim=-1)
+            XLora_scalings[nonzero_mask] = softmax_res_nonzero
+
+        return XLora_scalings
+
+
+class XLoraLinearLayer(XLoraLayer):
+    def __init__(
+        self,
+        model: nn.Module,
+        target: lora.Linear,
+        target_forward: Callable[..., Any],
+        layer_number: int,
+        config: XLoraConfig,
+    ) -> None:
+        super().__init__(model, target, target_forward, layer_number, config)
+
+    def forward(self, x: Tensor, *args: Any, **kwargs: Any) -> Tensor:
+        """
+        This method is designed to be a drop-in-replacement for the LoRA layers' .forward method. To use it, a bound
+        method must be created (bound to an instance of the XLoraLayer class).
+        """
+
+        previous_dtype = x.dtype
+        XLora_scalings = self.get_maybe_topk_scalings()
+
+        # Ignore if disabled. We want to make sure this is always run.
+        if self.target.merged:
+            result = self.target.base_layer(x, *args, **kwargs)
+        else:
+            result = self.target.base_layer(x, *args, **kwargs)
```

Comment:
No matter if the target is merged or not, we run the same code. I think that's correct, but it means we can remove the check.